sdks/python/apache_beam/runners/portability/fn_api_runner/trigger_manager.py

```diff
@@ -0,0 +1,459 @@
+        watermark_timer=watermark_timer,
+        current_time_state=None,
+        watermark_state=watermark_state,
+        elements_bag_state=element_bag,
+        window_tag_values_bag_state=window_tag_values,
+        finished_windows_state=finished_windows_state)
+    result = self._trigger_fire(
+        key, TimeDomain.WATERMARK, timestamp, timer_tag, context)
+    watermark_state.add(timestamp)
+    return result
+
+
+class FnRunnerStatefulTriggerContext(TriggerContext):
+  def __init__(
+      self,
+      processing_time_timer: RuntimeTimer,
+      watermark_timer: RuntimeTimer,
+      current_time_state: typing.Optional[AccumulatingRuntimeState],
+      watermark_state: typing.Optional[AccumulatingRuntimeState],
+      elements_bag_state: BagRuntimeState,
+      window_tag_values_bag_state: BagRuntimeState,
+      finished_windows_state: SetRuntimeState):
+    self.timers = {
+        TimeDomain.REAL_TIME: processing_time_timer,
+        TimeDomain.WATERMARK: watermark_timer
+    }
+    self.current_times = {
+        TimeDomain.REAL_TIME: current_time_state,
+        TimeDomain.WATERMARK: watermark_state
+    }
+    self.elements_bag_state = elements_bag_state
+    self.window_tag_values_bag_state = window_tag_values_bag_state
+    self.finished_windows_state = finished_windows_state
+
+  def windows_to_elements_map(
+      self
+  ) -> typing.Dict[BoundedWindow, typing.List[windowed_value.WindowedValue]]:
+    window_element_pairs: typing.Iterable[typing.Tuple[
+        BoundedWindow,
+        windowed_value.WindowedValue]] = self.elements_bag_state.read()
+    result: typing.Dict[BoundedWindow,
+                        typing.List[windowed_value.WindowedValue]] = {}
+    for w, e in window_element_pairs:
+      if w not in result:
+        result[w] = []
+      result[w].append(e)
+    return result
+
+  def for_window(self, window):
+    return PerWindowTriggerContext(window, self)
+
+  def get_current_time(self):
+    return self.current_times[TimeDomain.REAL_TIME].read()
+
+  def set_timer(self, name, time_domain, timestamp):
+    _LOGGER.debug('Setting timer (%s, %s) at %s', time_domain, name, timestamp)
+    self.timers[time_domain].set(timestamp, dynamic_timer_tag=name)
+
+  def clear_timer(self, name, time_domain):
+    _LOGGER.debug('Clearing timer (%s, %s)', time_domain, name)
+    self.timers[time_domain].clear(dynamic_timer_tag=name)
+
+  def merge_state(self, to_be_merged, merge_result):
+    all_triplets = self.window_tag_values_bag_state.read()
+    # Collect all the triplets for the window we are merging away, and tag them
+    # with the new window (merge_result).
+    merging_away_triplets = [(merge_result, t[1], t[2]) for t in all_triplets
+                             if t[0] in to_be_merged]
+
+    # Collect all of the other triplets, and joining them with the newly-tagged
+    # set of triplets.
+    resulting_triplets = [
+        t for t in all_triplets if t[0] not in to_be_merged
```

Comment:
just to confirm: is all_triplets reiterable?

Reply:
it is not reiterable